### PR TITLE
[usso] Adds usso auth support in `arc` tooling

### PR DESCRIPTION
**Pull request by artms**
Arc will first try working with domain without usso token (unless token cache for domain is not empty) and if call fails - it will try using ussh+usso combination to fetch uSSO token and save it to cache for later use.

### CODE REVIEW

File: src/workflow/ArcanistWorkflow.php [hunk 1]
```diff
@@ -2068,5 +2100,74 @@ private function getModernCommonDictionary(array $map) {
     return $map;
   }
 
+  // UBER CODE
+  // usso itself is somewhat sluggish, takes 1 second to return cached token
+  const USSO_CACHE_TIMEOUT = 600;
+
+  private static function getUSSOCacheFilename($domain) {
+    return implode(DIRECTORY_SEPARATOR,
+                   array(
```
Comment (Pawka) [inline src/workflow/ArcanistWorkflow.php:2109]: Indentation somehow broken.

File: src/workflow/ArcanistWorkflow.php [hunk 2]
```diff
@@ -2068,5 +2100,74 @@ private function getModernCommonDictionary(array $map) {
...
+                   sys_get_temp_dir(),
+                         sprintf('usso-token-cache-%s.json', md5($domain)),
+                   ));
+  }
+
+  private function maybeUseUSSOToken($domain) {
+    $cache = self::getUSSOCacheFilename($domain);
+    try {
+      $data = @json_decode(file_get_contents($cache), true);
```
Comment (Pawka) [inline src/workflow/ArcanistWorkflow.php:2118]: Hiding errors will make debugging difficult. Use only if necessary.

File: src/workflow/ArcanistWorkflow.php [hunk 3]
```diff
@@ -2068,5 +2100,74 @@ private function getModernCommonDictionary(array $map) {
...
+  }
+
+  private function getUSSOTokenForDomain($domain) {
+    // check ussh and if necessary ask to auth
+    list($e, $stdin, $stderr) = exec_manual('ussh');
+    if ($e != 0) {
+      $e = phutil_passthru('ussh');
+      if ($e != 0) {
+        throw new ArcanistUsageException(
+          pht('Looks like you missing `ussh` or it is not working'));
```
Comment (Pawka) [inline src/workflow/ArcanistWorkflow.php:2137]: `ussh` binary is missing or can not be executed.

File: src/workflow/ArcanistWorkflow.php [hunk 4]
```diff
@@ -2068,5 +2100,74 @@ private function getModernCommonDictionary(array $map) {
...
+      }
+    }
+    // try fetching usso token
+    list($e, $stdin, $stderr) = exec_manual('usso -ussh %s', $domain);
+    if ($e != 0) {
+      $e = phutil_passthru('usso -ussh %s', $domain);
+      if ($e != 0) {
+        throw new ArcanistUsageException(
+          pht('Looks like you missing `usso` or it is not working'));
```
Comment (Pawka) [inline src/workflow/ArcanistWorkflow.php:2146]: `usso` binary is missing or can not be executed.

File: src/workflow/ArcanistWorkflow.php [hunk 5]
```diff
@@ -2068,5 +2100,74 @@ private function getModernCommonDictionary(array $map) {
...
+    list($e, $stdin, $stderr) = exec_manual('usso -ussh %s -print', $domain);
+    if ($e != 0) {
+      throw new ArcanistUsageException($stdin."\n".$stderr);
+    }
+    $token = trim($stdin);
+    $old = umask(0077);
+    try {
+      $cache = self::getUSSOCacheFilename($domain);
+      file_put_contents($cache,
+                        json_encode(array(
```
Comment (Pawka) [inline src/workflow/ArcanistWorkflow.php:2159]: Indentation is broken.

Comment (artms) [reply src/workflow/ArcanistWorkflow.php]: Moved to separate file, fixed idendation

File: src/workflow/ArcanistWorkflow.php (hunk 2 above)
Comment (artms) [reply src/workflow/ArcanistWorkflow.php]: It is shortcut to check if file is readable and json is available - even if I check them only thing I can do is skip, as we discussed writing message that cache is empty doesn't make much sense for end user, in worst case user will see ~1second delay due to slow usso response

**Pull request merged**